src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+                        else -> {
+                          logger.log(Level.SEVERE, "Unsupported event timestamp type - ignoring")
+                          break
+                        }
+                      }
+                    val eventPath =
+                      EventPath(
+                        endpointId = endpoint.key.toUShort(),
+                        clusterId = cluster.key.toUInt(),
+                        eventId = eventList.key.toUInt()
+                      )
+
+                    val readData =
+                      ReadData.Event(
+                        path = eventPath,
+                        eventNumber = event.eventNumber.toULong(),
+                        priorityLevel = event.priorityLevel.toUByte(),
+                        timeStamp = timestamp,
+                        data = event.tlvValue
+                      )
+                    successes.add(readData)
+                  }
+                }
+              }
+            }
+          }
+
+          override fun onError(
+            attributePath: ChipAttributePath?,
+            eventPath: ChipEventPath?,
+            ex: Exception
+          ) {
+            attributePath?.let {
+              logger.log(Level.INFO, "Report error for attributePath:%s", it.toString())
+              val tmpAttributePath: AttributePath = attributePath.wrap()
+              val attributeFailure = ReadFailure.Attribute(path = tmpAttributePath, error = ex)
+              failures.add(attributeFailure)
+            }
+            eventPath?.let {
+              logger.log(Level.INFO, "Report error for eventPath:%s", it.toString())
+              val tmpEventPath: EventPath = eventPath.wrap()
+              val eventFailure = ReadFailure.Event(path = tmpEventPath, error = ex)
+              failures.add(eventFailure)
+            }
+
+            // The underlying subscription is terminated if both attributePath & eventPath are null
+            if (attributePath == null && eventPath == null) {
+              continuation.resumeWithException(
+                Exception("Read command failed with error ${ex.message}")
+              )
+            }
+          }
+
+          override fun onDone() {
+            logger.log(Level.FINE, "read command completed")
+            continuation.resume(ReadResponse(successes, failures))
+          }
+        }
+
+      deviceController.readPath(
+        reportCallback,
+        devicePtr,
+        chipAttributePaths,
+        chipEventPaths,
+        false,
+        CHIP_IM_TIMEOUT_MS,
+      )
+    }
+  }
+
+  override suspend fun write(writeRequests: WriteRequests): WriteResponse {
+    // To prevent potential issues related to concurrent modification, assign
+    // the value of the mutable property 'nodeId' to a temporary variable.
+    val nodeId = this.nodeId
+    check(nodeId != null) { "nodeId has not been initialized yet" }
+
+    val devicePtr: Long = getConnectedDevicePointer(nodeId)
```

Comment:
local copy of nodeId does not seem needed here.

Reply:
We do, same as above